jaxrs/src/main/java/com/facebook/airlift/jaxrs/thrift/ThriftMapper.java

```diff
@@ -0,0 +1,157 @@
+
+    @Inject
+    public ThriftMapper(ThriftCodecManager thriftCodecManager)
+    {
+        this.thriftCodecManager = thriftCodecManager;
+    }
+
+    @Override
+    public boolean isReadable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType)
+    {
+        return canReadOrWrite(type);
+    }
+
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType)
+    {
+        return canReadOrWrite(type);
+    }
+
+    private static boolean canReadOrWrite(Class<?> type)
+    {
+        if (IO_CLASSES.contains(type)) {
+            return false;
+        }
+        for (Class<?> ioClass : IO_CLASSES) {
+            if (ioClass.isAssignableFrom(type)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public Object readFrom(Class<Object> type,
+            Type genericType,
+            Annotation[] annotations,
+            MediaType mediaType,
+            MultivaluedMap<String, String> httpHeaders,
+            InputStream inputStream)
+            throws IOException
+    {
+        try {
+            ThriftCodec<?> codec = thriftCodecManager.getCodec(type);
+            TProtocol tProtocol = new TFacebookCompactProtocol(new TInputStreamTransport(inputStream));
+            return codec.read(tProtocol);
+        }
+        catch (Exception e) {
+            // we want to return a 400 for bad Thrift but not for a real IO exception
+            if (e instanceof IOException && !(e instanceof EOFException)) {
+                throw (IOException) e;
+            }
+            // log the exception at debug so it can be viewed during development
+            // Note: we are not logging at a higher level because this could cause a denial of service
+            log.debug(e, "Invalid Thrift for Java type %s", type);
+            // Invalid thrift request. Throwing exception so the response code can be overridden using a mapper.
+            throw new ThriftMapperParsingException(type, e);
+        }
+    }
+
+    @Override
+    public long getSize(Object value, Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType)
+    {
+        // In general figuring output size requires actual writing; usually not
+        // worth it to write everything twice.
+        return -1;
+    }
+
+    @Override
+    public void writeTo(Object value,
+            Class<?> type,
+            Type genericType,
+            Annotation[] annotations,
+            MediaType mediaType,
+            MultivaluedMap<String, Object> httpHeaders,
+            OutputStream outputStream)
+            throws IOException
+    {
+        try {
+            ThriftCodec codec = thriftCodecManager.getCodec(type);
+            TProtocol tProtocol = new TFacebookCompactProtocol(new TOutputStreamTransport(outputStream));
+            codec.write(value, tProtocol);
+        }
+        catch (Exception e) {
```

Comment:
Let's add a special handling for the `EOFException` (similar to read). Also take care of `TTransportException` wrapping.